File: app/src/main/java/com/thornbirds/repodemo/MainActivity.java
package com.thornbirds.repodemo;

import android.app.Fragment;
import android.app.FragmentManager;
import android.content.ComponentName;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import com.thornbirds.framework.activity.ComponentActivity;

public class MainActivity extends ComponentActivity {

    private Fragment mPersistFragment;
    private Fragment mAddFragment1;
    private Fragment mAddFragment2;
    private Fragment mReplaceFragment1;
    private Fragment mReplaceFragment2;

    private MyService.MyServiceImpl mMyService;

    private final ServiceConnection mServiceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            mMyService = (MyService.MyServiceImpl) service;
            Log.d(TAG, "onServiceDisconnected mMyService");
            mMyService.print("log onServiceConnected");
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            mMyService = null;
            Log.d(TAG, "onServiceDisconnected mMyService");
        }
    };

    @Override
    protected final String getTAG() {
        return "MainActivity";
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        FloatingActionButton fab = (FloatingActionButton) findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
                        .setAction("Action", null).show();
//                Intent intent = new Intent(MainActivity.this, com.thornbirds.framework.MainActivity.class);
//                startActivity(intent);
                FragmentManager fragmentManager = getFragmentManager();
                if (mAddFragment1 == null) {
                    Fragment fragment = new TestFragment();
                    Bundle bundle = new Bundle();
                    bundle.putInt(TestFragment.EXTRA_FRAGMENT_ID, 1);
                    fragment.setArguments(bundle);
                    fragmentManager.beginTransaction().add(R.id.root_container, fragment).commit();
                    mAddFragment1 = fragment;
                } else if (mAddFragment2 == null) {
                    Fragment fragment = new TestFragment();
                    Bundle bundle = new Bundle();
                    bundle.putInt(TestFragment.EXTRA_FRAGMENT_ID, 2);
                    fragment.setArguments(bundle);
                    fragmentManager.beginTransaction().add(R.id.root_container, fragment).commit();
                    mAddFragment2 = fragment;
                } else if (mReplaceFragment1 == null) {
                    Fragment fragment = new TestFragment();
                    Bundle bundle = new Bundle();
                    bundle.putInt(TestFragment.EXTRA_FRAGMENT_ID, 3);
                    fragment.setArguments(bundle);
                    fragmentManager.beginTransaction().replace(R.id.root_container, fragment).commit();
                    mReplaceFragment1 = fragment;
                } else if (mReplaceFragment2 == null) {
                    Fragment fragment = new TestFragment();
                    Bundle bundle = new Bundle();
                    bundle.putInt(TestFragment.EXTRA_FRAGMENT_ID, 4);
                    fragment.setArguments(bundle);
                    fragmentManager.beginTransaction().replace(R.id.root_container, fragment).commit();
                    mReplaceFragment2 = fragment;
                }
            }
        });

        mPersistFragment = getFragmentManager().findFragmentById(R.id.test_fragment);

        final Intent intent = new Intent(this, MyService.class);
        bindService(intent, mServiceConnection, BIND_AUTO_CREATE);
    }

    @Override
    protected void onResume() {
        super.onResume();
        Log.d(TAG, "onResume mMyService");
        if (mMyService != null) {
            mMyService.print("log onResume");
        }
    }

    @Override
    protected void onStop() {
        super.onStop();
        Log.d(TAG, "onStop mMyService");
        if (mMyService != null) {
            mMyService.print("log onStop");
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Log.d(TAG, "onDestroy mMyService");
        if (mMyService != null) {
            mMyService.print("log onDestroy");
        }
        unbindService(mServiceConnection);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/thornbirds/repodemo/MainActivity.kt
package com.thornbirds.repodemo

import android.content.ComponentName
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.snackbar.Snackbar
import com.thornbirds.framework.activity.ComponentActivity
import com.thornbirds.repodemo.MyService.MyServiceImpl

class MainActivity : ComponentActivity() {

    override val TAG: String = "MainActivity"

    private var mPersistFragment: Fragment? = null
    private var mAddFragment1: Fragment? = null
    private var mAddFragment2: Fragment? = null
    private var mReplaceFragment1: Fragment? = null
    private var mReplaceFragment2: Fragment? = null
    private var mMyService: MyServiceImpl? = null
    private val mServiceConnection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            mMyService = service as MyServiceImpl
            Log.d(TAG, "onServiceDisconnected mMyService")
            mMyService!!.print("log onServiceConnected")
        }

        override fun onServiceDisconnected(name: ComponentName) {
            mMyService = null
            Log.d(TAG, "onServiceDisconnected mMyService")
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val toolbar = findViewById<View>(R.id.toolbar) as Toolbar
        setSupportActionBar(toolbar)
        val fab = findViewById<View>(R.id.fab) as FloatingActionButton
        fab.setOnClickListener { view ->
            Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
                .setAction("Action", null).show()
            //                Intent intent = new Intent(MainActivity.this, com.thornbirds.framework.MainActivity.class);
            //                startActivity(intent);
            val fragmentManager =
                supportFragmentManager
            if (mAddFragment1 == null) {
                val fragment: Fragment = TestFragment()
                val bundle = Bundle()
                bundle.putInt(TestFragment.EXTRA_FRAGMENT_ID, 1)
                fragment.arguments = bundle
                fragmentManager.beginTransaction().add(R.id.root_container, fragment).commit()
                mAddFragment1 = fragment
            } else if (mAddFragment2 == null) {
                val fragment: Fragment = TestFragment()
                val bundle = Bundle()
                bundle.putInt(TestFragment.EXTRA_FRAGMENT_ID, 2)
                fragment.arguments = bundle
                fragmentManager.beginTransaction().add(R.id.root_container, fragment).commit()
                mAddFragment2 = fragment
            } else if (mReplaceFragment1 == null) {
                val fragment: Fragment = TestFragment()
                val bundle = Bundle()
                bundle.putInt(TestFragment.EXTRA_FRAGMENT_ID, 3)
                fragment.arguments = bundle
                fragmentManager.beginTransaction().replace(R.id.root_container, fragment).commit()
                mReplaceFragment1 = fragment
            } else if (mReplaceFragment2 == null) {
                val fragment: Fragment = TestFragment()
                val bundle = Bundle()
                bundle.putInt(TestFragment.EXTRA_FRAGMENT_ID, 4)
                fragment.arguments = bundle
                fragmentManager.beginTransaction().replace(R.id.root_container, fragment).commit()
                mReplaceFragment2 = fragment
            }
        }
        mPersistFragment = supportFragmentManager.findFragmentById(R.id.test_fragment)
        val intent = Intent(this, MyService::class.java)
        bindService(intent, mServiceConnection, BIND_AUTO_CREATE)
    }

    override fun onResume() {
        super.onResume()
        Log.d(TAG, "onResume mMyService")
        if (mMyService != null) {
            mMyService!!.print("log onResume")
        }
    }

    override fun onStop() {
        super.onStop()
        Log.d(TAG, "onStop mMyService")
        if (mMyService != null) {
            mMyService!!.print("log onStop")
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        Log.d(TAG, "onDestroy mMyService")
        if (mMyService != null) {
            mMyService!!.print("log onDestroy")
        }
        unbindService(mServiceConnection)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        val id = item.itemId
        return if (id == R.id.action_settings) {
            true
        } else super.onOptionsItemSelected(item)
    }
}